src/proper_erlang_abstract_code.erl

```diff
@@ -0,0 +1,3175 @@
+
+singleton_integer_type(S) ->
+    ?LET(It,
+         wunion([int, char, unop, binop], S, abstract_type),
+         begin
+             It
+         end).
+
+small(S, pattern) ->
+    wunion([atom, bool, int, string, char, float, nil, pat_var],
+           S, ?FUNCTION_NAME);
+small(S, _Where) ->
+    wunion([atom, bool, int, string, char, float, nil, var],
+           S, ?FUNCTION_NAME).
+
+a_bool(_S) ->
+    proper_types:union([lit_atom('true'), lit_atom('false')]).
+
+an_integer(_S) ->
+    lit_integer(proper_types:non_neg_integer()).
+
+a_string(S) ->
+    {'string', anno(), simple_string(S)}.
+
+simple_string(S) ->
+    N = uniform(get_limit(string, S) + 1) - 1,
+    simple_string1(S, N).
+
+simple_string1(_S, 0) ->
+    [];
+simple_string1(S, N) ->
+    [simple_char(S) | simple_string1(S, N - 1)].
+
+a_char(S) ->
+    {'char', anno(), simple_char(S)}.
+
+simple_char(S) ->
+    ?LET(C,
+         (S#gen_state.simple_char)(),
+         begin
+             C
+         end).
+
+default_simple_char() ->
+    proper_types:union([proper_types:integer($a, $z),
+                        proper_types:integer($A, $Z)]).
+
+default_atom() ->
+    any_of(some_atoms()).
+
+a_float(_S) ->
+    ?LET(Float,
+         proper_types:float(),
+         {'float', anno(), abs(Float)}).
+
+var(_S) ->
+    a_variable(bound_var_or_an_atom).
+
+pat_var(S) ->
+    wunion([fresh_var, bound_var], S, ?FUNCTION_NAME).
+
+fresh_var(S) ->
+    case get_weight(fresh_var, S) of
+        0 ->
+            anonymous_var(S);
+        _ ->
+            proper_types:weighted_union(
+              [{20, ?LAZY(a_variable(fresh_var))},
+               {1, ?LAZY(anonymous_var(S))}])
+    end.
+
+bound_var(_S) ->
+    a_variable(bound_var_or_an_integer).
+
+anonymous_var(_S) ->
+    a_variable('_').
+
+a_variable(Name) ->
+    {'var', anno(), Name}.
+
+match(S, pattern) ->
+    {'match', anno(), pattern(S), pattern(S)};
+match(S, compound) ->
+    {'match', anno(), pattern(S), abstract_expr(S)}.
+
+tuple(S, abstract_type) ->
+    tuple_type(S);
+tuple(S, Where) ->
+    T = where(S, Where),
+    {'tuple', anno(), list_of_gen(T, get_limit(tuple, S))}.
+
+non_empty_list_of_gen(G, Max) ->
+    N = uniform(Max),
+    list_of_gen2(N, G).
+
+list_of_gen(G, Max) ->
+    N = uniform(Max + 1) - 1,
+    list_of_gen2(N, G).
+
+list_of_gen2(0, _G) ->
+    [];
+list_of_gen2(N, G) ->
+    [G | list_of_gen2(N - 1, G)].
+
+bitstring(S, abstract_type) ->
+    {'type', anno(), 'binary', [singleton_integer_type(S),
+                                singleton_integer_type(S)]};
+bitstring(S, Where) ->
+    case S#gen_state.result_type of
+        term ->
+            wunion([bits, bytes], S, Where);
+        ResType when ResType =:= program;
+                     ResType =:= guard;
+                     ResType =:= expr  ->
+            wunion([bits, blc], S, Where)
+    end.
+
+bytes(S) ->
+    {'bin', anno(), binelement_seq_term(S, bytes)}.
+
+bits(S, compound) when S#gen_state.result_type =:= term ->
```

Comment:
Suggestion by Tidier:
```erlang
bits(#gen_state{result_type = term} = S, compound) ->
```

Alternatively, factor the first two clauses with `compound` as 2nd argument and use a case switch in the body, to make it clear that these are the cases for compound.